app/src/main/java/com/mamadou/newsapp/MainActivity.kt

```diff
@@ -3,54 +3,75 @@ import android.content.Context
-import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
+import androidx.lifecycle.lifecycleScope
 import com.mamadou.newsapp.databinding.ActivityMainBinding
 import com.mamadou.newsapp.networking.NetworkStatusChecker
 import com.mamadou.newsapp.networking.RemoteApi
 import com.mamadou.newsapp.networking.buildApiService
-import java.util.*
+import com.mamadou.newsapp.utils.CustomResult
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.Dispatchers.Main
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
 
 class MainActivity : AppCompatActivity() {
 
     private lateinit var binding: ActivityMainBinding
-    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
     private val networkStatusChecker by lazy {
         NetworkStatusChecker(this.getSystemService(Context.CONNECTIVITY_SERVICE)  as ConnectivityManager)
     }
 
+    private  val service = buildApiService()
+    private  val remoteApi = RemoteApi(service)
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
         binding = ActivityMainBinding.inflate(layoutInflater)
-
         setContentView(binding.root)
+        binding.articleRecyclerView.run {
+            adapter = articleAdapter
+        }
+        fetchArticles()
+        binding.swiperefresh.setOnRefreshListener {
+            fetchArticles()
+        }
+    }
+
+    private val articleAdapter =
+        ArticleRecyclerAdapter { article ->
+            val newsDetailIntent = Intent(this@MainActivity, NewsDetailsActivity::class.java)
+            newsDetailIntent.putExtra(INTENT_EXTRA_ARTICLE, article)
+            startActivity(newsDetailIntent)
+        }
 
+
+    private fun fetchArticles() {
         networkStatusChecker.performIfConnectedToInternet ({
-            val service = buildApiService()
-            val remoteApi = RemoteApi(service)
-            remoteApi.getArticles { articles, error ->
-                if (error != null) {
-                    Log.e("There is an error", error.toString())
-
-                } else {
-                    binding.articleRecyclerView.run {
-                        adapter = ArticleRecyclerAdapter(articles) { articleIndex ->
-                            val newsDetailIntent = Intent(this@MainActivity, NewsDetailsActivity::class.java)
-                            newsDetailIntent.putExtra(INTENT_EXTRA_ARTICLE, articles[articleIndex])
-                            startActivity(newsDetailIntent)
+            binding.swiperefresh.isRefreshing = true
+            lifecycleScope.launch(IO) {
+                val result = remoteApi.getArticles()
+                withContext(Main) {
+                    when(result) {
+                        is CustomResult.Success -> {
+                                articleAdapter.updateArticle(result.value)
+                        }
+                        is CustomResult.Failure -> {
                         }
```

Comment:
Make sure to wire this Failure case up to an error Toast or dialog!

Reply:
Toast method created!